assets/javascripts/discourse/lib/ai-streamer.js

```diff
@@ -41,69 +42,140 @@ export function addProgressDot(element) {
+        `#post_${this.post.post_number}`
+      );
     }
-    return true;
   }
 
-  if (!post) {
-    // wait till later
-    return false;
+  get element() {
+    return this.postElement;
   }
 
-  const oldRaw = post.get("raw") || "";
+  get loaded() {
+    return !!this.postElement;
+  }
 
-  if (postStatus.raw === oldRaw && !postStatus.done) {
-    const hasProgressDot =
-      postElement && postElement.querySelector(".progress-dot");
-    if (hasProgressDot) {
-      return false;
+  set streaming(value) {
+    if (this.postElement) {
+      if (value) {
+        this.postElement.classList.add("streaming");
+      } else {
+        this.postElement.classList.remove("streaming");
+      }
     }
   }
 
-  if (postStatus.raw) {
-    const newRaw = postStatus.raw.substring(
-      0,
-      oldRaw.length + LETTERS_PER_INTERVAL
-    );
-    const cooked = await cook(newRaw);
-
-    post.set("raw", newRaw);
-    post.set("cooked", cooked);
+  async setRaw(value, done) {
+    this.post.set("raw", value);
+    const cooked = await cook(value);
 
     // resets animation
-    postElement.classList.remove("streaming");
-    void postElement.offsetWidth;
-    postElement.classList.add("streaming");
+    this.element.classList.remove("streaming");
+    void this.element.offsetWidth;
+    this.element.classList.add("streaming");
 
     const cookedElement = document.createElement("div");
     cookedElement.innerHTML = cooked;
 
-    addProgressDot(cookedElement);
+    if (!done) {
+      addProgressDot(cookedElement);
+    }
+
+    await this.setCooked(cookedElement.innerHTML);
+  }
+
+  async setCooked(value) {
+    this.post.set("cooked", value);
 
-    const element = document.querySelector(
-      `#post_${postStatus.post_number} .cooked`
-    );
+    const oldElement = this.postElement.querySelector(".cooked");
 
     await loadScript("/javascripts/diffhtml.min.js");
-    window.diff.innerHTML(element, cookedElement.innerHTML);
+    window.diff.innerHTML(oldElement, value);
   }
 
-  if (postStatus.done) {
-    if (postElement) {
-      postElement.classList.remove("streaming");
+  get raw() {
+    return this.post.get("raw") || "";
+  }
+}
+
+export async function applyProgress(status, updater) {
+  status.startTime = status.startTime || Date.now();
+
+  if (Date.now() - status.startTime > GIVE_UP_INTERVAL) {
+    updater.streaming = false;
+    return true;
+  }
+
+  if (!updater.loaded) {
+    // wait till later
+    return false;
+  }
+
+  const oldRaw = updater.raw;
+
+  if (status.raw === oldRaw && !status.done) {
+    const hasProgressDot = updater.element.querySelector(".progress-dot");
+    if (hasProgressDot) {
+      return false;
     }
   }
 
-  return postStatus.done;
+  if (status.raw !== undefined) {
+    let newRaw = status.raw;
+
+    if (!status.done) {
+      // rush update if we have a </details> tag (function call)
+      if (oldRaw.length === 0 && newRaw.indexOf("</details>") !== -1) {
```

Comment:
```suggestion
      if (oldRaw.length === 0 && newRaw.includes("</details>")) {
```

Reply:
will change but I need to cover this code